pkg/cache/cluster.go

```diff
@@ -1192,3 +1230,89 @@ func (c *clusterCache) GetClusterInfo() ClusterInfo {
 func skipAppRequeuing(key kube.ResourceKey) bool {
 	return ignoredRefreshResources[key.Group+"/"+key.Kind]
 }
+
+// UpdateClusterConnectionStatus starts a goroutine that checks for watch failures.
+// If there are any watch errors, it will periodically ping the remote cluster
+// and update the cluster connection status.
+func (c *clusterCache) UpdateClusterConnectionStatus(ctx context.Context) {
+	go c.clusterConnectionService(ctx)
+}
+
+func (c *clusterCache) clusterConnectionService(ctx context.Context) {
+	clusterConnectionTimeout := 10 * time.Second
+	ticker := time.NewTicker(clusterConnectionTimeout)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			watchErrors := c.watchFails.len()
+			// Ping the cluster for connection verification if there are watch failures or
+			// if the cluster has recovered back from watch failures.
+			if watchErrors > 0 || (watchErrors == 0 && c.connectionStatus == ConnectionStatusFailed) {
+				c.log.V(1).Info("verifying cluster connection", "watches", watchErrors)
+
+				_, err := c.kubectl.GetServerVersion(c.config)
+				if err != nil {
+					if c.connectionStatus != ConnectionStatusFailed {
+						c.updateConnectionStatus(ConnectionStatusFailed)
+					}
+				} else if c.connectionStatus != ConnectionStatusSuccessful {
+					c.updateConnectionStatus(ConnectionStatusSuccessful)
+				}
+			}
+		case <-ctx.Done():
+			ticker.Stop()
+			return
+		}
+	}
+
+}
```

Comment:
I've refactored the function with locks to read the c.connectionStatus. Managed to avoid locking around `GetServerVersion`. Hopefully, this should fix the issue.